src/Xrefcheck/Scanners/Markdown.hs

```diff
@@ -82,33 +84,73 @@ cataNode f (Node pos ty subs) = f pos ty (cataNode f <$> subs)
 
 -- | Remove nodes with accordance with global `MarkdownConfig` and local
 --   overrides.
-removeIgnored :: Node -> Either Text Node
-removeIgnored = runIdentity . runExceptT . flip evalStateT None . cataNode remove
+removeIgnored :: MonadWriter [ScanError] m => FilePath -> Node -> m Node
+removeIgnored fp = withIgnoreMode . cataNode remove
   where
     remove
-      :: (MonadError Text m, MonadState IgnoreMode m)
+      :: (MonadWriter [ScanError] m, MonadState IgnoreMode m)
       => Maybe PosInfo
       -> NodeType
       -> [m Node]
       -> m Node
     remove pos ty subs = do
       mode <- get
+      let node = Node pos ty []
       case (mode, ty) of
         (Paragraph, PARAGRAPH) -> put None $> defNode
-        (Paragraph, x)         -> throwError (makeError mode (prettyType x) pos)
-        (File, _)              -> throwError (makeError mode "" pos)
-        (Link, LINK {})        -> put None $> defNode
-        (Link, _)              -> Node pos ty <$> sequence subs
-        (None, _) -> do
-          case getIgnoreMode (Node pos ty []) of
-            Just mode' -> put mode' $> defNode
+        (Paragraph, x)         -> do
+          tell . scanError pos . makeError mode $ prettyType x
+          put None
+          Node pos ty <$> sequence subs
+        (File, _)              -> do
+          tell . scanError pos $ makeError mode ""
+          put None
+          Node pos ty <$> sequence subs
+        (Link _, LINK {})      -> put None $> defNode
+        (Link pos', _)         ->
+          case getIgnoreMode node of
+            Just mode'  -> do
+              tell . scanError pos' $ makeError mode ""
+              handleMode node mode'
+            Nothing     -> Node pos ty <$> sequence subs
```

Comment:
Can you please explain this `(Link pos', _)` case?

If the mode is currently `Link` and we find a non-LINK node, shouldn't we unconditionally report a scan error? Why do we check `getIgnoreMode` and why does the `Nothing` branch not report an error?

Reply:
One of the examples in the `README` shows that the link does not have to follow the annotation
https://github.com/serokell/xrefcheck/blob/c4f0bfb9fc771d706902b9741f3b3aaca5ac47c7/README.md?plain=1#L102-L106
Same for tests 
https://github.com/serokell/xrefcheck/blob/f4eb187f31918367f71b741a3066f0632070d394/tests/markdowns/with-annotations/ignore_link.md?plain=1#L34-L38
I thought it was on purpose... :thinking:  I tried to keep the old behavior. Do I need to change this?

Reply:
> Here we have text node before the link

> I tried to keep the old behavior

Oh you're absolutely right, I forgot about that!

Ok, can you sprinkle some comments in this function to explain all these decisions? E.g.
* In the `Paragraph` case, we should explain that "When we find an `ignore paragraph` annotation, we expect to find a paragraph immediately after"
* In the `Link` case, explain that when we find an `ignore link` annotation, we skip nodes until (1) we find a link, (2) find another `ignore` annotation, in which case we throw an error, or (3) reach the end of the file, in which case we also throw an error.
* We don't expect to find an `ignore file` annotation at all while traversing the nodes. We've **already** used `checkIgnoreFile` to extract those annotations located at the top of the file, and handle them. If we do find one anyway, that means it's **not** at the top of the file, in which case we throw an error.

You don't necessarily have to use my words, these are just examples of things I think are not completely obvious from a first glance and would benefit from better docs. Feel free to add more notes as well if you think they could be useful.

Also, please add some haddocks to `checkIgnoreFile` and `getIgnoreMode` while we're here :D

Reply:
Great job!